language/bytecode-verifier/src/dependencies.rs

```diff
@@ -1,415 +1,341 @@
-
-        // verify dependencies
-        checker.verify_imported_modules(module.module_handles(), Some(module.self_handle_idx()))?;
-        checker.verify_imported_structs(module.struct_handles(), Some(module.self_handle_idx()))?;
-        checker.verify_imported_functions(module.function_handles(), Some(module.self_handle_idx()))
-    }
-
-    pub fn verify_script(
+    fn script(
         script: &'a CompiledScript,
-        dependencies: impl IntoIterator<Item = &'a CompiledModule>,
-    ) -> VMResult<()> {
-        Self::verify_script_impl(script, dependencies).map_err(|e| e.finish(Location::Script))
+        dependencies: impl IntoIterator<Item = &'b CompiledModule>,
+    ) -> Self {
+        Self::new(BinaryIndexedView::Script(script), dependencies)
     }
 
-    pub fn verify_script_impl(
-        script: &'a CompiledScript,
-        dependencies: impl IntoIterator<Item = &'a CompiledModule>,
-    ) -> PartialVMResult<()> {
-        let mut dependency_map = BTreeMap::new();
-        for dependency in dependencies {
-            let dependency_id = dependency.self_id();
-            dependency_map.insert(dependency_id, dependency);
-        }
+    fn new(
+        resolver: BinaryIndexedView<'a>,
+        dependencies: impl IntoIterator<Item = &'b CompiledModule>,
+    ) -> Self {
+        let self_module = resolver.self_id();
+        let dependency_map = dependencies
+            .into_iter()
+            .filter(|d| Some(d.self_id()) != self_module)
+            .map(|d| (d.self_id(), d))
+            .collect();
+
         let mut checker = Self {
-            resolver: BinaryIndexedView::Script(script),
+            resolver,
             dependency_map,
             struct_id_to_handle_map: HashMap::new(),
             func_id_to_handle_map: HashMap::new(),
         };
-        checker.build_deps_entry_point();
-
-        checker.verify_imported_modules(script.module_handles(), None)?;
-        checker.verify_imported_structs(script.struct_handles(), None)?;
-        checker.verify_imported_functions(script.function_handles(), None)
-    }
 
-    fn build_deps_entry_point(&mut self) {
-        for (module_id, module) in &self.dependency_map {
+        for (module_id, module) in &checker.dependency_map {
             // Module::StructName -> def handle idx
             for struct_def in module.struct_defs() {
                 let struct_handle = module.struct_handle_at(struct_def.struct_handle);
                 let struct_name = module.identifier_at(struct_handle.name);
-                self.struct_id_to_handle_map.insert(
+                checker.struct_id_to_handle_map.insert(
                     (module_id.clone(), struct_name.to_owned()),
                     struct_def.struct_handle,
                 );
             }
             // Module::FuncName -> def handle idx
             for func_def in module.function_defs() {
-                if !func_def.is_public {
+                if !func_def.is_public() {
```

Comment:
I guess this might conflict with #7418 as well.

Reply:
It will, yeah